rust/lance-encoding/src/encodings/logical/primitive.rs

```diff
@@ -563,7 +584,374 @@ impl FieldEncoder for PrimitiveFieldEncoder {
+
+            let num_value_bytes = chunk.num_bytes as usize;
+            let values =
+                &miniblocks.data[value_offset as usize..value_offset as usize + num_value_bytes];
+            data_buffer.extend_from_slice(values);
+
+            data_buffer.extend(iter::repeat(0).take(pad_bytes as usize));
+
+            value_offset += num_value_bytes as u64;
+        }
+
+        (
+            LanceBuffer::Owned(data_buffer),
+            LanceBuffer::Owned(meta_buffer),
+        )
+    }
+
+    /// Compresses a buffer of levels
+    ///
+    /// TODO: Use bit-packing here
+    fn compress_levels(
+        levels: Option<LevelBuffer>,
+        num_values: u64,
+        compression_strategy: &dyn CompressionStrategy,
+        chunks: &[MiniBlockChunk],
+    ) -> Result<(Vec<LanceBuffer>, pb::ArrayEncoding)> {
+        if let Some(levels) = levels {
+            debug_assert_eq!(num_values as usize, levels.len());
+            // Make the levels into a FixedWidth data block
+            let mut levels_buf = LanceBuffer::reinterpret_vec(levels);
+            let levels_block = DataBlock::FixedWidth(FixedWidthDataBlock {
+                data: levels_buf.borrow_and_clone(),
+                bits_per_value: 16,
+                num_values,
+            });
+            let levels_field = Field::new_arrow("", DataType::UInt16, false)?;
+            // Pick a block compressor
+            let (compressor, compressor_desc) =
+                compression_strategy.create_block_compressor(&levels_field, &levels_block)?;
+            // Compress blocks of levels (sized according to the chunks)
+            let mut buffers = Vec::with_capacity(chunks.len());
+            let mut off = 0;
+            let mut values_counter = 0;
+            for chunk in chunks {
+                let chunk_num_values = chunk.num_values(values_counter, num_values);
+                values_counter += chunk_num_values;
+                let level_bytes = chunk_num_values as usize * 2;
+                let chunk_levels = levels_buf.slice_with_length(off, level_bytes);
+                let chunk_levels_block = DataBlock::FixedWidth(FixedWidthDataBlock {
+                    data: chunk_levels,
+                    bits_per_value: 16,
+                    num_values: chunk_num_values,
+                });
+                let compressed_levels = compressor.compress(chunk_levels_block)?;
+                off += level_bytes;
+                buffers.push(compressed_levels);
+            }
+            Ok((buffers, compressor_desc))
+        } else {
+            // Everything is valid or we have no repetition so we encode as a constant
+            // array of 0
+            let data = chunks.iter().map(|_| LanceBuffer::empty()).collect();
+            let scalar = 0_u16.to_le_bytes().to_vec();
+            let encoding = ProtobufUtils::constant(scalar, num_values);
+            Ok((data, encoding))
+        }
+    }
+
+    fn encode_miniblock(
```

Comment:
we may need to adjust the function signature here later when we want to do recursively encoding(`arrays` -> `datablock`, `Result<EncodedPage>` ->`Result<DataBlock>`) but I think it is fine now

Reply:
I think only `MiniBlockCompressor::compress` will need recursion, not `encode_miniblock` but I do agree we still need to figure out that recursion still.